WEB-INTEGRATION.md

```diff
@@ -0,0 +1,742 @@
+algorithm, and [“fire an event”](https://dom.spec.whatwg.org/#concept-event-fire)
+which wraps it, will create the event with dispatchSnapshot set to null. These
+algorithms will be extended with an optional boolean argument
+_**populateDispatchSnapshot**_ (default false) that will cause
+`dispatchSnapshot` to instead be set to the current AsyncContext mapping.
+[\[DOM\]](https://dom.spec.whatwg.org/)
+
+For events with a sync dispatch context, setting this argument to true will be
+enough. For events with an asynchronous dispatch context, that context will have
+to be set as the current context first:
+
+> 1. Let _context_ be [AsyncContextSnapshot](https://tc39.es/proposal-async-context/#sec-asynccontextsnapshot)().
+> 1. [Queue a global task](https://html.spec.whatwg.org/multipage/webappapis.html#queue-a-global-task)
+>    to run the following steps:
+>    1. Run the AsyncContext Snapshot _context_ while performing the following
+>       steps:
+>       1. [Fire an event](https://dom.spec.whatwg.org/#concept-event-fire)
+>          named `foo`, with _**populateDispatchSnapshot**_ set to true, at
+>          _target_.
+
+## Runtime script errors
+
+One important part of the way AsyncContext works is strict stacking – you can
+only switch into a context by calling a function, which restores the previous
+context when it returns or throws. This, however, would make exposing the throw
+context in the error event impossible, since the spec text that calls into JS
+would not have a way to get that throw context, if it is different from the
+context that the spec text set for that JS execution.
+
+To allow the error event to have the throw context,
+[PR #95](https://github.com/tc39/proposal-async-context/pull/95) on the
+AsyncContext proposal adds a `[[ThrowAsyncContextMapping]]` field on the agent
+record. This field might still be set to `EMPTY` even when an exception was
+thrown, but in those cases the throw context will be the same as the current
+context.
+
+The `error` event is fired through the
+[“report an error”](https://html.spec.whatwg.org/multipage/webappapis.html#report-the-error)
+and [“report an exception”](https://html.spec.whatwg.org/multipage/webappapis.html#report-the-exception)
+spec algorithms [\[HTML\]](https://html.spec.whatwg.org/multipage/), which are
+usually called immediately after a script execution or callback invocation.
+Those algorithms will be updated to use `[[ThrowAsyncContextMapping]]` to get
+the throw context, but in order to have the right context even when it is
+`EMPTY`, they must be called before the call to `AsyncContextSwap()` that exits
+the context in which the script is run:
+
+> 1. Let _context_ be [AsyncContextSnapshot](https://tc39.es/proposal-async-context/#sec-asynccontextsnapshot)().
+> 1. [Queue a global task](https://html.spec.whatwg.org/multipage/webappapis.html#queue-a-global-task)
+>    to run the following steps:
+>    1. Run the AsyncContext Snapshot _context_ while performing the following
```

Comment:
Why is this necessary? Why can't Web IDL callback integration take care of this for web spec authors?